src/index.js

```diff
@@ -70,42 +69,241 @@ const connectMQTT = async availabilityTopic => {
+    //
+    //      Why not use a switch?  Because buttons/events better represent the paradigm presented
+    //      by onstar.  A switch is a state that can be monitored, but onstar doesn't allow us
+    //      to check if the car is currently running to turn off the "startVehicle" switch. We can
+    //      only send the command (button) and see the immediate response (event.)
+    //
+    //      The buttons and events will have to have similar, but different, names.  For example,
+    //      the button would be startVehicle and the event would be startVehicle-result. (I'm not 
+    //      thrilled with that. Can an autodiscovery config topic have payloads for different 
+    //      types of entities?  The state topics would absolutely have to be different... and that
+    //      would really break the paradigm in mqtt.js.)
+    //
+    //      ... and it turns out that MQTT-events are broken.  That's fine.  To keep things easier
+    //      for automations, just use a single "command_result" event.  Within that event, stuff
+    //      all kinds of data, including the command sent, the success/failure state, a timestamp,
+    //      a friendly name, and even a friendly message that can be directly copied to a notification.    
 
+    const buttonCommands = [ 
+        { cmd: 'startVehicle', friendlyName: "Start", icon: "mdi:car-key" }, 
+        { cmd: 'cancelStartVehicle', friendlyName: "Cancel Start", icon: "mdi:car-off" }, 
+        { cmd: 'alert', friendlyName: "Flash Lights and Honk Horn", icon: "mdi:alarm-light" }, 
+        { cmd: 'alertFlash', friendlyName: "Flash Lights", icon: "mdi:alarm-light" }, 
+        { cmd: 'alertHonk', friendlyName: "Honk Horn", icon: "mdi:alarm-light" }, 
+        { cmd: 'lockDoor', friendlyName: "Lock Doors", icon: "mdi:car-door-lock" }, 
+        { cmd: 'unlockDoor', friendlyName: "Unlock Doors", icon: "mdi:car-door" }, 
+        { cmd: 'getLocation', friendlyName: "Get Location", icon: "mdi:map-marker", noevent: true }
+    ];
+
+    // map it to an array of objects to make things easier
+
+    const button_discovery = [];
+    buttonCommands.forEach(cmd => {
+        const [ ctopic, cpayload ] = mqttHA.getButtonConfig(cmd.cmd);
+        if (cmd.icon)
+            cpayload.icon = cmd.icon;
+        if (cmd.friendlyName)
+            cpayload.name = MQTT.convertFriendlyName(cpayload.name);
+
+        logger.info('topic: ' + ctopic);
+        logger.info('payload: ' + JSON.stringify(cpayload));
+        button_discovery.push(
+            client.publish(ctopic, JSON.stringify(cpayload), {retain: true})
+        );
+        // just publish a single comment event and stick all the results in there
+        const [ etopic, epayload ] = mqttHA.getEventConfig('command_result');
+        button_discovery.push(
+            client.publish(etopic, JSON.stringify(epayload), {retain: true}));
+        
+        Promise.all(button_discovery);
```

Comment:
I'm not at all familiar with javascript async code and just followed the model that was used elsewhere. ;)  I was under the impression (only from context) that "Promise.all(list)" would ensure everything in the "list" was completed.  Would it be better to just call client.publish() multiple times without using promises?

Reply:
The all() does that, but nothing is using the one promise to represent the many. If a publish were to fail (without a `.catch()`), it may throw an unhandled rejection exception and crash. Not really the biggest deal here, docker will restart the service and these actions are mostly fire and forget.

Removing the promise collection and .all() would be fine.